### PR TITLE
Introduce Metadata for documents

### PR DESCRIPTION
**Pull request by shumkov**

### CODE REVIEW

File: lib/document/Document.js [hunk 1]
```diff
@@ -165,20 +170,73 @@ class Document {
...
+    const json = {
       $type: this.getType(),
       $scope: this.scope,
       $scopeId: this.scopeId,
       $rev: this.getRevision(),
       $action: this.getAction(),
       ...this.getData(),
     };
+
+    if (this.meta) {
```
Comment (shumkov) [inline lib/document/Document.js]: Is it work with an empty object? Probably better with `Object.keys(this.meta).length > 0`?

File: test/integration/document/validateDocumentFactory.spec.js [hunk 2]
```diff
@@ -339,6 +339,21 @@ describe('validateDocumentFactory', () => {
         expect(error.getRawDocument()).to.equal(rawDocument);
       });
     });
+
```
Comment (shumkov) [inline test/integration/document/validateDocumentFactory.spec.js]: We need to test property names too

File: lib/document/enrichContractWithBaseDocument.js [hunk 3]
```diff
@@ -5,7 +5,7 @@ const documentBaseSchema = require('../../schema/base/document');
  * @param {Contract} contract
  * @return {RawContract}
  */
-function enrichContractWithBaseDocument(contract) {
+function enrichContractWithBaseDocument(contract, excludeFromRequired = ['$meta']) {
```
Comment (shumkov) [inline lib/document/enrichContractWithBaseDocument.js]: excludeBaseProprety = []

File: lib/document/validateDocumentFactory.js [hunk 4]
```diff
@@ -67,6 +68,14 @@ module.exports = function validateDocumentFactory(
 
       const enrichedContract = enrichContractWithBaseDocument(contract);
```
Comment (shumkov) [inline lib/document/validateDocumentFactory.js]: `enrichContractWithBaseDocument(contract, options.allowMeta ? [] : ['$meta']);`

File: lib/document/enrichContractWithBaseDocument.js [hunk 5]
```diff
@@ -26,7 +26,9 @@ function enrichContractWithBaseDocument(contract) {
```
Comment (shumkov) [inline lib/document/enrichContractWithBaseDocument.js]: Clone and update (`properties`, `required`) base document according to `excludeBaseProprety `

File: lib/document/Document.js [hunk 6]
```diff
@@ -7,6 +7,8 @@ const DataIsNotAllowedWithActionDeleteError = require('./errors/DataIsNotAllowed
 const hash = require('../util/hash');
 const { encode } = require('../util/serializer');
 
+const Meta = require('./Meta');
```
Comment (shumkov) [inline lib/document/Document.js]: Let's call it `Metadata`?

File: lib/document/Document.js [hunk 7]
```diff
@@ -41,6 +43,11 @@ class Document {
       delete data.$rev;
     }
 
+    if (Object.prototype.hasOwnProperty.call(rawDocument, '$meta')) {
+      this.meta = new Meta(rawDocument.$meta.userId);
```
Comment (shumkov) [inline lib/document/Document.js]: Let's pass the whole `rawDocument.$meta`. The document class should not know about meta internals.

File: lib/document/Document.js [hunk 8]
```diff
@@ -165,29 +172,63 @@ class Document {
...
+  }
+
+  /**
+   * Remove metadate from document
+   *
+   * @return {Document}
+   */
+  removeMetadata() {
+    this.meta = undefined;
+    return this;
```
Comment (shumkov) [inline lib/document/Document.js]: newline ;)

File: lib/document/Document.js [hunk 9]
```diff
@@ -165,29 +172,63 @@ class Document {
...
+      json.$meta = this.meta.toJSON();
+    }
+
+    return json;
   }
 
   /**
    * Return serialized Document
    *
+   * @param {Object} options
```
Comment (shumkov) [inline lib/document/Document.js]: Mark as optional `[ ]`

File: lib/document/Document.js [hunk 10]
```diff
@@ -165,29 +172,63 @@ class Document {
...
+   * @param {boolean} options.skipMeta
```
Comment (shumkov) [inline lib/document/Document.js]: Mark as optional and set default value `[options.skipMeta=false]`

File: lib/document/Meta.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
+class Meta {
+  /**
+   * Create document meta class
+   * @param {string} userId
+   */
+  constructor(userId) {
+    this.userid = userId;
```
Comment (shumkov) [inline lib/document/Meta.js]: `userId`

File: lib/document/Meta.js [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  /**
+   * Get user id
```
Comment (shumkov) [inline lib/document/Meta.js]: Get user ID. Newline after description

File: lib/document/RawDocumentInterface.js [hunk 13]
```diff
@@ -5,4 +5,5 @@
  * @property {string} $scopeId
  * @property {number} $rev
  * @property {number} $action
+ * @property {RawDocumentMeta|undefined} $meta
```
Comment (shumkov) [inline lib/document/RawDocumentInterface.js]: I would keep these names consistent. Rename this one to `RawMetadata` or `Metdata` to `DocumentMetadata`.

File: lib/document/enrichContractWithBaseDocument.js [hunk 14]
```diff
@@ -18,16 +19,19 @@ function enrichContractWithBaseDocument(contract) {
   Object.keys(clonedDocuments).forEach((type) => {
     const clonedDocument = clonedDocuments[type];
 
-    const { properties: baseDocumentProperties } = documentBaseSchema;
+    const { properties: baseDocumentProperties, required: baseRequired } = documentBaseSchema;
```
Comment (shumkov) [inline lib/document/enrichContractWithBaseDocument.js]: Let's keep names consistent: `baseDocumentRequired`.

File: lib/document/enrichContractWithBaseDocument.js [hunk 15]
```diff
@@ -3,9 +3,10 @@ const documentBaseSchema = require('../../schema/base/document');
 /**
  * @typedef {enrichContractWithBaseDocument}
  * @param {Contract} contract
+ * @param {string[]} excludeBaseProperty
  * @return {RawContract}
  */
-function enrichContractWithBaseDocument(contract) {
+function enrichContractWithBaseDocument(contract, excludeBaseProperty = []) {
```
Comment (shumkov) [inline lib/document/enrichContractWithBaseDocument.js]: `excludeBaseDocumentProperties`

File: lib/document/enrichContractWithBaseDocument.js [hunk 16]
```diff
@@ -18,16 +19,19 @@ function enrichContractWithBaseDocument(contract) {
...
-      clonedDocument.properties[name] = baseDocumentProperties[name];
-      clonedDocument.required.push(name);
-    });
+    Object.keys(baseDocumentProperties)
+      .filter(name => excludeBaseProperty.indexOf(name) === -1)
+      .forEach((name) => {
+        clonedDocument.properties[name] = baseDocumentProperties[name];
+      });
+
+    baseRequired.forEach(name => clonedDocument.required.push(name));
```
Comment (shumkov) [inline lib/document/enrichContractWithBaseDocument.js]: If `required` is not defined in clonedDocument?

File: lib/document/validateDocumentFactory.js [hunk 17]
```diff
@@ -24,9 +24,10 @@ module.exports = function validateDocumentFactory(
    * @typedef validateDocument
    * @param {Document|RawDocument} document
    * @param {Contract} contract
+   * @param {Object} [options={allowMeta: true}]
```
Comment (shumkov) [inline lib/document/validateDocumentFactory.js]: Let's make as above:
```
* @param {Object} [options]
* @param {Object} [options.allowMeta=true]
```

File: schema/base/document.json [hunk 18]
```diff
@@ -28,6 +28,13 @@
       "type": "string",
       "minLength": 34,
       "maxLength": 34
+    },
+    "$meta": {
```
Comment (shumkov) [inline schema/base/document.json]: We decided to hardcode meta fields, so you can remove `propertyNames` and define `userId` in `properties` and `additionalProperties: false`.

File: test/unit/document/Document.spec.js [hunk 19]
```diff
@@ -286,6 +286,36 @@ describe('Document', () => {
...
+  describe('#getMetadata', () => {
+    it('should return all meta', () => {
+      const meta = {
+        test1: 1,
+        test2: 2,
+      };
+
+      document.meta = meta;
+
+      expect(document.getMetadata()).to.equal(meta);
```
Comment (shumkov) [inline test/unit/document/Document.spec.js]: Here should be `DocumentMetadata` object?

File: lib/document/enrichContractWithBaseDocument.js (hunk 16 above)
Comment (jawid-h) [reply lib/document/enrichContractWithBaseDocument.js]: This logic is happening a little bit earlier.
```
if (!clonedDocument.required) {
  clonedDocument.required = [];
}
```

**Pull request merged**